src/schnorr.rs

```diff
@@ -54,21 +54,21 @@ impl Secp256k1 {
             return Err(Error::IncapableContext)
         }
 
-        let mut ret: Signature = unsafe { mem::uninitialized() };
         unsafe {
+            let mut ret = mem::MaybeUninit::<Signature>::uninit();
             // We can assume the return value because it's not possible to construct
             // an invalid signature from a valid `Message` and `SecretKey`
             let err = ffi::secp256k1_schnorr_sign(
                 self.ctx,
-                ret.as_mut_ptr(),
+                (*ret.as_mut_ptr()).as_mut_ptr(),
```

Comment:
What is the effect of this expression `(*ret.as_mut_ptr()).as_mut_ptr()`?
Could you explain it to me?

Reply:
Yes, it looks weird, but it has a reason.

Pointers don't have a Deref trait unlikely references, because it's not safe when the pointer is null. So you cannot call methods through a pointer, while you can call them through a reference. The safe way to use a pointer is using [`as_ref`](https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref) that returns `Option`. The result will be `None` if the pointer is a null pointer. However, when you ensure the pointer is not null, you can dereference the pointer directly. Of course, it needs an `unsafe` block because the compiler cannot know whether it's safe or not.

And there is no way to get a mutable reference of the underlying type from `MaybeUninit`. The [method](https://doc.rust-lang.org/beta/std/mem/union.MaybeUninit.html#method.get_mut) to get a reference is not stable yet. So I used [`MaybeUninit::<T>::as_mut_ptr()`](https://doc.rust-lang.org/beta/std/mem/union.MaybeUninit.html#method.as_mut_ptr) and dereference the pointer directly.

Reply:
Thanks! I understood it.